File: Zeze/Raft/RocksRaft/Changes.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Zeze.Serialize;

namespace Zeze.Raft.RocksRaft
{
	public class Changes : Zeze.Raft.Log
	{
		// 收集日志时,记录所有Bean修改.
		// key is Bean.ObjectId
		public Dictionary<long, LogBean> Beans { get; } = new Dictionary<long, LogBean>();
		public Rocks Rocks { get; }
		public Transaction Transaction { get; }
		public Dictionary<int, long> AtomicLongs { get; } = new Dictionary<int, long>();

		public Changes(Rocks r, Transaction t, Zeze.Net.Protocol req)
			: base((IRaftRpc)req)
        {
			Rocks = r;
			Transaction = t;
		}

		internal Changes(Rocks r)
			: base(null)
        {
			Rocks = r;
        }

		public class Record
        {
			public const int Remove = 0;
			public const int Put = 1;
			public const int Edit = 2;

			public int TableTemplateId { get; set; }
			public string TableTemplateName { get; set; }

			public int State { get; set; }
			public Bean PutValue { get; set; }
			public ISet<LogBean> LogBean { get; } = new HashSet<LogBean>();

			// 所有的日志修改树，key is Record.Value。不会被系列化。
			public Util.IdentityHashMap<Bean, LogBean> LogBeans { get; } = new Util.IdentityHashMap<Bean, LogBean>();

			public Table Table;

			public void SetTableByName(Changes changes, string templateName)
            {
				Table = changes.Rocks.GetTableTemplate(templateName).OpenTable(TableTemplateId);
			}

			public void Collect(Transaction.RecordAccessed ar)
            {
				if (null != ar.PutLog) // put or remove
				{
					PutValue = ar.PutLog.Value;
					State = PutValue == null ? Remove : Put;
					return;
				}

				State = Edit;
				if (LogBeans.TryGetValue(ar.Origin.Value, out var logbean))
					LogBean.Add(logbean); // edit
			}

			public void Encode(ByteBuffer bb)
            {
				bb.WriteUInt(State);
				switch (State)
				{
					case Remove:
						break;

					case Put:
						PutValue.Encode(bb);
						break;

					case Edit:
						bb.Encode(LogBean);
						break;
				}
			}

			public void Decode(ByteBuffer bb)
            {
				State = bb.ReadUInt();
				switch (State)
                {
					case Remove:
						break;

					case Put:
						PutValue = Table.NewValue();
						PutValue.Decode(bb);
						break;

					case Edit:
						bb.Decode(LogBean);
						break;
                }
            }

            public override string ToString()
            {
				var sb = new StringBuilder();

				sb.Append("State=").Append(State);
				sb.Append(" PutValue=").Append(PutValue);
				sb.Append("\nLog=");
				ByteBuffer.BuildString(sb, LogBean);
				sb.Append("\nAllLog=");
				ByteBuffer.BuildString(sb, LogBeans.Values);

				return sb.ToString();
            }
        }

		// 收集记录的修改,以后需要系列化传输.
		public Dictionary<TableKey, Record> Records { get; } = new Dictionary<TableKey, Record>();

		public void Collect(Bean recent, Log log)
		{
			if (null == log.Belong)
            {
				// 记录可能存在多个修改日志树。收集的时候全部保留，后面会去掉不需要的。see Transaction._final_commit_
				if (false == Records.TryGetValue(recent.TableKey, out var r))
				{
					r = new Record()
					{
						TableTemplateId = recent.RootInfo.Record.Table.TemplateId,
						TableTemplateName = recent.RootInfo.Record.Table.TemplateName,
					};
					Records.Add(recent.TableKey, r);
				}
				r.LogBeans.TryAdd(recent, (LogBean)log);
				return; // root
			}

			if (false == Beans.TryGetValue(log.Belong.ObjectId, out LogBean logbean))
			{
				if (log.Belong is Collection)
				{
					// 容器使用共享的日志。需要先去查询，没有的话才创建。
					logbean = (LogBean)Transaction.Current.GetLog(log.Belong.Parent.ObjectId + log.Belong.VariableId);
				}
				if (null == logbean)
                {
					logbean = log.Belong.CreateLogBean();
				}
				Beans.Add(log.Belong.ObjectId, logbean);
			}
			logbean.Collect(this, log.Belong, log);
		}

		public void CollectRecord(Transaction.RecordAccessed ar)
        {
			var tkey = ar.TableKey;

			if (false == Records.TryGetValue(tkey, out var r))
            {
				// put record only
				r = new Record()
				{
					TableTemplateId = ar.Origin.Table.TemplateId,
					TableTemplateName = ar.Origin.Table.TemplateName,
				};
				Records.Add(tkey, r);
			}

			r.Collect(ar);
		}

		public override void Decode(ByteBuffer bb)
		{
			for (int i = bb.ReadUInt(); i > 0; i--)
			{
				var tkey = new TableKey();
                var r = new Record
                {
                    TableTemplateId = bb.ReadUInt(),
                    TableTemplateName = bb.ReadString()
                };

                tkey.Name = bb.ReadString();
				r.SetTableByName(this, r.TableTemplateName);
				r.Table.DecodeKey(bb, out tkey.Key);

				r.Decode(bb);

				Records.Add(tkey, r);
			}

			for (int i = bb.ReadUInt(); i > 0; --i)
			{
				var index = bb.ReadUInt();
				var value = bb.ReadLong();
				AtomicLongs.Add(index, value);
			}
		}

		public override void Encode(ByteBuffer bb)
		{
			bb.WriteUInt(Records.Count);
			foreach (var r in Records)
			{
				// encode TableTemplate
				bb.WriteUInt(r.Value.TableTemplateId);
				bb.WriteString(r.Value.TableTemplateName);

				// encode TableKey
				r.Value.SetTableByName(this, r.Value.TableTemplateName);
				bb.WriteString(r.Key.Name);
				r.Value.Table.EncodeKey(bb, r.Key.Key);

				// encode record
				r.Value.Encode(bb);
			}
			bb.WriteUInt(AtomicLongs.Count);
			foreach (var a in AtomicLongs)
            {
				bb.WriteUInt(a.Key);
				bb.WriteLong(a.Value);
            }
		}

        public override string ToString()
        {
			var sb = new StringBuilder();
			ByteBuffer.BuildString(sb, Records);
            return sb.ToString();
        }

        public override async Task Apply(RaftLog holder, StateMachine stateMachine)
        {
			if (holder.LeaderFuture != null)
			{
				Rocks.logger.Debug($"{Rocks.Raft.Name} LeaderApply");
				await Transaction.LeaderApply(this);
			}
            else
            {
				Rocks.logger.Debug($"{Rocks.Raft.Name} FollowerApply");
				await Rocks.FollowerApply(this);
			}
		}
	}
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Raft/RocksRaft/Changes.java
package Zeze.Raft.RocksRaft;

import java.util.HashMap;
import java.util.HashSet;
import java.util.IdentityHashMap;
import java.util.Set;
import Zeze.Raft.IRaftRpc;
import Zeze.Raft.RaftLog;
import Zeze.Raft.StateMachine;
import Zeze.Serialize.ByteBuffer;
import Zeze.Serialize.IByteBuffer;
import Zeze.Util.IntHashMap;
import Zeze.Util.LongHashMap;

public final class Changes extends Zeze.Raft.Log {
	private final Rocks rocks;
	private final LongHashMap<LogBean> beans = new LongHashMap<>(); // 收集日志时,记录所有Bean修改. key is Bean.ObjectId
	private final HashMap<TableKey, Record> records = new HashMap<>(); // 收集记录的修改,以后需要序列化传输.
	private final IntHashMap<Long> atomicLongs = new IntHashMap<>();
	private Transaction transaction;
	public static final int TypeId_ = Zeze.Transaction.Bean.hash32(Changes.class.getName());

	@Override
	public long typeId() {
		return TypeId_;
	}

	public Changes(Rocks r) {
		super(null);
		rocks = r;
	}

	public Changes(Rocks r, Transaction t, IRaftRpc req) {
		super(req);
		rocks = r;
		transaction = t;
	}

	public Rocks getRocks() {
		return rocks;
	}

	public LongHashMap<LogBean> getBeans() {
		return beans;
	}

	public HashMap<TableKey, Record> getRecords() {
		return records;
	}

	public IntHashMap<Long> getAtomicLongs() {
		return atomicLongs;
	}

	public static final class Record {
		public static final int Remove = 0;
		public static final int Put = 1;
		public static final int Edit = 2;

		private int tableTemplateId;
		private String tableTemplateName;
		private int state;
		private Bean putValue;
		private final Set<LogBean> logBean = new HashSet<>();
		// 所有的日志修改树，key is Record.Value。不会被序列化。
		private final IdentityHashMap<Bean, LogBean> logBeans = new IdentityHashMap<>();
		public Table<?, ?> table;

		public void setTableTemplateId(int value) {
			tableTemplateId = value;
		}

		public String getTableTemplateName() {
			return tableTemplateName;
		}

		public void setTableTemplateName(String value) {
			tableTemplateName = value;
		}

		public int getState() {
			return state;
		}

		public Bean getPutValue() {
			return putValue;
		}

		public Set<LogBean> getLogBean() {
			return logBean;
		}

		public IdentityHashMap<Bean, LogBean> getLogBeans() {
			return logBeans;
		}

		public void setTableByName(Changes changes, String templateName) {
			var tableTpl = changes.rocks.getTableTemplate(templateName);
			if (tableTpl == null) {
				var names = new StringBuilder();
				for (String name : changes.rocks.getTableTemplates().keySet())
					names.append(' ').append(name);
				throw new IllegalStateException("unknown table template: " + templateName + ", available:" + names);
			}
			table = tableTpl.openTable(tableTemplateId);
		}

		public void collect(Transaction.RecordAccessed ar) {
			if (ar.getPutLog() != null) { // put or remove
				putValue = ar.getPutLog().value;
				state = putValue == null ? Remove : Put;
				return;
			}

			state = Edit;
			var logBean = logBeans.get(ar.getOrigin().getValue());
			if (logBean != null)
				this.logBean.add(logBean); // edit
		}

		public void encode(ByteBuffer bb) {
			bb.WriteInt(state);
			switch (state) {
			case Remove:
				break;
			case Put:
				putValue.encode(bb);
				break;
			case Edit:
				bb.encode(logBean);
				break;
			}
		}

		public void decode(IByteBuffer bb) {
			state = bb.ReadInt();
			switch (state) {
			case Remove:
				break;
			case Put:
				putValue = table.newValue();
				putValue.decode(bb);
				break;
			case Edit:
				bb.decode(logBean, LogBean::new);
				break;
			}
		}

		@Override
		public String toString() {
			var sb = new StringBuilder();
			sb.append("State=").append(state).append(" PutValue=").append(putValue);
			sb.append("\nLog=");
			ByteBuffer.BuildSortedString(sb, logBean);
			sb.append("\nAllLog=");
			ByteBuffer.BuildSortedString(sb, logBeans.values());
			return sb.toString();
		}
	}

	public void collect(Bean recent, Log log) {
		Bean belong = log.getBelong();
		if (belong == null) {
			// 记录可能存在多个修改日志树。收集的时候全部保留，后面会去掉不需要的。see Transaction._final_commit_
			var r = records.get(recent.tableKey());
			if (r == null) {
				r = new Record();
				var table = recent.rootInfo().getRecord().getTable();
				r.setTableTemplateId(table.getTemplateId());
				r.setTableTemplateName(table.getTemplateName());
				records.put(recent.tableKey(), r);
			}
			r.getLogBeans().put(recent, (LogBean)log);
			return; // root
		}

		var logBean = beans.get(belong.objectId());
		if (logBean == null) {
			if (belong instanceof Collection) {
				// 容器使用共享的日志。需要先去查询，没有的话才创建。
				logBean = (LogBean)Transaction.getCurrent().getLog(
						belong.parent().objectId() + belong.variableId());
			}
			if (logBean == null)
				logBean = belong.createLogBean();
			beans.put(belong.objectId(), logBean);
		}
		logBean.collect(this, belong, log);
	}

	public void collectRecord(Transaction.RecordAccessed ar) {
		var tkey = ar.tableKey();
		var r = records.get(tkey);
		if (r == null) {
			// put record only
			r = new Record();
			var table = ar.getOrigin().getTable();
			r.setTableTemplateId(table.getTemplateId());
			r.setTableTemplateName(table.getTemplateName());
			records.put(tkey, r);
		}

		r.collect(ar);
	}

	@Override
	public void apply(RaftLog holder, StateMachine stateMachine) {
		if (holder.isLeaderRequest()) {
			if (Rocks.isDebugEnabled)
				Rocks.logger.debug("{} LeaderApply", rocks.getRaft().getName());
			transaction.leaderApply(this);
		} else {
			// Rocks.logger.debug("{} followerApply", rocks.getRaft().getName());
			rocks.followerApply(this);
		}
	}

	@SuppressWarnings("unchecked")
	@Override
	public void encode(ByteBuffer bb) {
		bb.WriteUInt(records.size());
		for (var r : records.entrySet()) {
			// encode TableTemplate
			bb.WriteUInt(r.getValue().tableTemplateId);
			bb.WriteString(r.getValue().tableTemplateName);

			// encode TableKey
			r.getValue().setTableByName(this, r.getValue().tableTemplateName);
			bb.WriteString(r.getKey().name);
			((Table<Object, Bean>)r.getValue().table).encodeKey(bb, r.getKey().key);

			// encode record
			r.getValue().encode(bb);
		}
		bb.WriteUInt(atomicLongs.size());
		for (var it = atomicLongs.iterator(); it.moveToNext(); ) {
			bb.WriteUInt(it.key());
			bb.WriteLong(it.value());
		}
	}

	@Override
	public void decode(IByteBuffer bb) {
		for (int i = bb.ReadUInt(); i > 0; i--) {
			var tkey = new TableKey();
			var r = new Record();

			r.setTableTemplateId(bb.ReadUInt());
			r.setTableTemplateName(bb.ReadString());

			tkey.name = bb.ReadString();
			r.setTableByName(this, r.getTableTemplateName());
			tkey.key = r.table.decodeKey(bb);

			r.decode(bb);

			records.put(tkey, r);
		}

		for (int i = bb.ReadUInt(); i > 0; i--) {
			var index = bb.ReadUInt();
			var value = bb.ReadLong();
			atomicLongs.put(index, value);
		}
	}

	@Override
	public String toString() {
		var sb = new StringBuilder();
		ByteBuffer.BuildString(sb, records);
		return sb.toString();
	}
}
